Frictional functions for mechanical contact

recuero opened the pull request:
Closes #21024

recuero committed:
Initial commit (#21024)

Use actual (and not weighted) relative velocity to compute the frictional coefficient
from a function.

recuero committed:
Add simple frictional model to quasi static mortar frictional contstraints (#21024)

For quasistatic, we assume we can add the _current_ contact pressure and _current_
relative velocity into the frictional function without affecting convergence too much.
This will need to be tested in practical problems.

recuero committed:
Change BC to get parallel consistency (#21024)

Somehow driving the cube via a sinusoidal function was causing a parallel issue.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1050170/) on ff1e184 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b7a3ce</th>
      <th colspan=4>#21050 ff1e18</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.57%</td>
      <td>82.57%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>75204</td>
      <td>75203</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15878</td>
      <td>15879</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21050/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21050/coverage/framework)


## Modules coverage

### Contact

<table>
  <thead>
    <tr>
      <th></th>
      <th>b7a3ce</th>
      <th colspan=4>#21050 ff1e18</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.01%</td>
      <td>89.00%</td>
      <td>-0.01%</td>
      <td>95.65%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>2876</td>
      <td>2913</td>
      <td>+37</td>
      <td>66</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>355</td>
      <td>360</td>
      <td>+5</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21050/coverage_diff/contact)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21050/coverage/contact)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21050/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad committed:
Catch MetaPhysicL exception from mortar constraints

Refs #20175

lindsayad committed:
Only request solution states in sync

lindsayad committed:
Allow solution state addition from displaced mesh objects

lindsayad committed:
Pre-request old solution state in constructor

Antonio was getting hangs (before my `libmesh_parallel_only`
assertion addition) because one process was requesting the old
solution vector while another process was not. And then when
we were updating our system vectors at timestep end, we would
be trying to scatter ghosts from one vector on one process and
a different vector on another process, which lead to the hang.
By requesting the old solution state in sync on all processes,
we avoid that issue

recuero committed:
Remove maps in favor of solutionOld()

lindsayad commented on modules/contact/src/constraints/ComputeFrictionalForceLMMechanicalContact.C:
```diff
@@ -29,13 +29,16 @@ ComputeFrictionalForceLMMechanicalContact::validParams()
       1.0e-7,
       "Minimum value of contact pressure that will trigger frictional enforcement");
-  params.addRequiredParam<Real>("mu", "The friction coefficient for the Coulomb friction law");
+  params.addParam<Real>("mu", 0.5, "The friction coefficient for the Coulomb friction law");
```
why this default choice?

recuero replied:
The code in the constructor and residual computation made it such a default value was never used. A default for friction values is misleading I think. So I am switching to a range checked parameter without a default value to be more consistent code-wise.

recuero committed:
Avoid providing unnecessary default value for coefficient of friction.

Addressing Alex's comment.
